### PR TITLE
Reconcile the Java DSL Source with Flow

### LINKED ISSUE #25472: Audit required between the Java Flow and Source FlowOps?
Having just played at the corners of the Java DSL for streams today, I made two quick discoveries of new functions having been applied to `Flow` but not `Source`. Unlike the Scala DSL and its `FlowOps` it seems that the Java DSL has a requirement for duplicating functions in two places.

Would it be worth conducting an audit between the Java DSL `Flow` and `Source` to ensure that they are consistent?

Discoveries:

* https://github.com/akka/akka/pull/25036/files#r209869320
* https://github.com/akka/akka/pull/24984#issuecomment-412705382

### PR DESCRIPTION
**Pull request by huntc**
For the Java DSL, Flow had operations that were not in Source.

Performed a manual reconciliation. Several functions were missing.

Fixes #25472

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: This was not a typo but a 'math-ism' (https://en.wikipedia.org/wiki/If_and_only_if). I like user simpler language though. Perhaps really spell out: "Terminate processing (and cancel the upstream publisher) after predicate returns false for the first time. When 'inclusive' is true, also include the element for which the predicate returned 'false'."

**Comment by raboof** [comment]
OK TO TEST

**Comment by raboof** [comment]
Did we find out why DslConsistencySpec didn't catch those?

**Comment by raboof** [comment]
Might be good to add code examples (#25468) to check the API's can be invoked as intended?

**Comment by huntc** [comment]
That was completely lost on me! Also, I don’t think the Source description of the equivalent had iff. I don’t have it in front of me though.

**Comment by huntc** [comment]
Yes, it sounds like a good idea. I observed this myself when reading the docs just today!

Can we relegate that to a separate PR though? I want to address the problem at hand with immediacy.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (huntc) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: That was completely lost on me! It certainly looks like a typo to us mere mortals.

**Comment by huntc** [comment]
@raboof Do you require anything further from me on this?

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (raboof) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Jup! Could you either revert back to iff or, even better, to a more clear (but verbose) description like I proposed?

**Review by raboof — COMMENTED** [review]
Seems almost good to go, might be nice to roll back (or forward) on the iff comment before merging.

**Review by ktoso — APPROVED** [review]
LGTM once @raboof's comment is addressed; Thanks Christopher

**Comment by huntc** [comment]
@raboof Restored the `Flow` based descriptions.

**Review by raboof — APPROVED** [review]

**Comment by raboof** [comment]
I moved the 'iff' improvement suggestion to https://github.com/akka/akka/issues/25486. Would be great if you could provide code examples (#25468) to check the API's can be invoked as intended in a subsequent PR.

Thanks!

**Review by ktoso — APPROVED** [review]
LGTM, thanks

**Comment by huntc** [comment]
Thanks guys. Ready to merge?

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala
Comment (pvlugter) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Just got this deprecation in some tests against akka snapshot. Should this have been on `recoverWith` methods instead? Given that the deprecation in the scaladsl is on `recoverWith` not `recover`:

https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala#L698

Comment (pvlugter) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: I'd say the one that was copied was a mistake. The other deprecations are on `recoverWith`. Original PR: #20137

Comment (huntc) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Here's the `Source` method showing the deprecation from 2.5.14:

https://github.com/akka/akka/blob/v2.5.14/akka-stream/src/main/scala/akka/stream/javadsl/Source.scala#L1103

I see that the Scala DSL has no such deprecation, so may be the deprecations on `recover` are misplaced in general?

Comment (huntc) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: So perhaps here's the original culprit: https://github.com/akka/akka/pull/20137/files#diff-dd7989ad8864b9e69daa0ddc6619e9efR821

Comment (pvlugter) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Yeah, that's the original misplaced deprecation.

Comment (huntc) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Perhaps @patriknw could confirm where the deprecations should be given his original PR. I'm happy to make another change.

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: Created https://github.com/akka/akka/issues/25531 to track

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: And concurrently Patrik created #25530 :D

**Pull request merged**
